content/blog/2021/06/kwctl-intro-for-kubernetes-administrators.md

```diff
@@ -0,0 +1,277 @@
+
+```shell
+$ kwctl run registry://ghcr.io/kubewarden/policies/psp-capabilities:v0.1.4 \
+        -r req_pod_with_container_with_capabilities_added.json \
+        --settings-json '{"default_add_capabilities": ["AUDIT_READ"], "allowed_capabilities": ["AUDIT_READ", "NET_ADMIN", "SYS_TIME"]}'
+```
+
+This time the policy accepts the incoming request, plus it mutates it.
+This is the output returned:
+
+```json
+{
+  "uid": "1299d386-525b-4032-98ae-1949f69f9cfc",
+  "allowed": true,
+  "patchType": "JSONPatch",
+  "patch": "W3sib3AiOiJhZGQiLCJwYXRoIjoiL3NwZWMvY29udGFpbmVycy8wL3NlY3VyaXR5Q29udGV4dC9jYXBhYmlsaXRpZXMvYWRkLzIiLCJ2YWx1ZSI6IkFVRElUX1JFQUQifV0="
+}
+```
+
+Alice can see what the policy is going to change by looking at the contents of
+the `patch` attribute.
+As requested by Kubernetes, the patch operation is encoded using `base64`:
+
+```shell
+$ kwctl run registry://ghcr.io/kubewarden/policies/psp-capabilities:v0.1.4 \
+        -r req_pod_with_container_with_capabilities_added.json \
+        --settings-json '{"default_add_capabilities": ["AUDIT_READ"], "allowed_capabilities": ["AUDIT_READ", "NET_ADMIN", "SYS_TIME"]}' \
+  | jq -r .patch | base64 -d
+```
+
+This produces the following output:
+```json
+[{"op":"add","path":"/spec/containers/0/securityContext/capabilities/add/2","value":"AUDIT_READ"}]
+```
+
+The policy is working as expected!
+
+## Deploying the policy
+
+After more experiments with `kwctl run`, Alice wants to deploy the policy
+inside of a Kubernetes cluster.
+
+Kubewarden policies are defined using the
+[ClusterAdmissionPolicy](https://github.com/kubewarden/kubewarden-controller/blob/main/docs/crds/README.asciidoc)
+Custom Resource provided by Kubewarden.
+
+Alice can quickly scaffold the YAML file defining this resource using kwctl:
+
+```shell
+$ kwctl manifest -t ClusterAdmissionPolicy registry://ghcr.io/kubewarden/policies/psp-capabilities:v0.1.4
+```
+
+The command will produce the following output:
+
+```yaml
+apiVersion: policies.kubewarden.io/v1alpha2
+kind: ClusterAdmissionPolicy
+metadata:
+  name: generated-policy
+spec:
+  module: "registry://ghcr.io/kubewarden/policies/psp-capabilities:v0.1.4"
+  settings: {}
+  rules:
+    - apiGroups:
+        - ""
+      apiVersions:
+        - v1
+      resources:
+        - pods
+      operations:
+        - CREATE
+  mutating: true
+```
+
+kwctl just saved quite some time to Alice 😊
+
+## Mirroring policies
+
+Kubewarden policies are distributed via OCI container registries, the very
+same pieces of infrastructure already used to distribute container images.
+
+Alice doesn't want her cluster to pull the Kubewarden policy straight from
+`ghcr.io/kubewarden/policies/psp-capabilities:v0.1.4`, she rather prefers to
+fetch it from a local registry she controls.
+
+This can be done using the following commands:
+
+```shell
+$ kwctl tag registry://ghcr.io/kubewarden/policies/psp-capabilities:v0.1.4 registry://registry.alice.corp.lan/policies/psp-capabilities:v0.1.4
```

Comment:
I think `tag` might need a bit more thinking. On a first thought I think I prefer `push` to be able to accept `file://`, `registry://` and `https://` as a source, instead of forcefully having to go through the main kubewarden store by tagging. I can understand this middle step in Docker/Podman for example, where images are multi layered, and layers are in the store. For Kubewarden I don't see a very strong point in having to `tag` before `push`.

In this instance, this would be:

`kwctl push registry://ghcr.io/kubewarden/policies/psp-capabilities:v0.1.4  registry://registry.alice.corp.lan/policies/psp-capabilities:v0.1.4`

A regular push as we have been doing until now would be:

`kwctl push file:///home/user/policy.wasm registry://registry.alice.corp.lan/policies/psp-capabilities:v0.1.4`

or the short form, as `run` does (assuming `/home/user` as the `PWD` in this case):

`kwctl push policy.wasm registry://registry.alice.corp.lan/policies/psp-capabilities:v0.1.4`

Reply:
This https://github.com/kubewarden/kwctl/issues/41 tracks the changes we have to implement. I'll update the blog post accordingly